Tests/text-formatting.lisp

```diff
@@ -0,0 +1,70 @@
+                (loop for (edge value) on spec-1 by #'cddr
+                      unless (equal value (getf spec-2 edge))
+                        do
+                           (return nil)
+                      finally
+                         (return t)))))
+    (is (equals (climi::normalize-margin-spec nil
+                                              '(:left (:relative 0)
+                                                :top (:relative 0)
+                                                :right (:relative 0)
+                                                :bottom (:relative 0)))
+                '(:left (:relative 0)
+                  :top (:relative 0)
+                  :right (:relative 0)
+                  :bottom (:relative 0))))
+    (is (equals (climi::normalize-margin-spec '(:left 42
+                                                :top 42
+                                                :right (:absolute 3)
+                                                :bottom 0)
+                                              '(:left (:relative 1)
+                                                :top (:relative 1)
+                                                :right (:relative 1)
+                                                :bottom (:relative 1)))
+                '(:left (:relative 42)
+                  :top (:relative 42)
+                  :right (:absolute 3)
+                  :bottom (:relative 0))))
+    (is (equals (climi::normalize-margin-spec '(:left (:absolute 1)
+                                                :top (:relative 1)
+                                                :bottom 0)
+                                              '(:left (:absolute 1)
+                                                :top (:absolute 1)
+                                                :right (:absolute 1)
+                                                :bottom (:absolute 1)))
+                '(:left (:absolute 1)
+                  :top (:relative 1)
+                  :right (:absolute 1)
+                  :bottom (:relative 0))))
+    ;; spec validity
+    (let ((default '(:left (:relative 0)
+                     :top (:relative 0)
+                     :right (:relative 0)
+                     :bottom (:relative 0))))
+      
+      (signals error
+        (climi::normalize-margin-spec
+         `(,(5am:gen-one-element :left :right :top :bottom)
+           (,(5am:gen-one-element :relative :absolute) :foo))
+         default))
+      (signals error
+        (climi::normalize-margin-spec
+         `(,(5am:gen-one-element :left :right :top :bottom)
+           :foo)
+         default))
+      (signals error
+        (climi::normalize-margin-spec
+         `(,(5am:gen-one-element :left :right :top :bottom)
+           (:relative 1 2))
+         default))
+      (signals error
+        (climi::normalize-margin-spec `(:left* (:relative 42))
+                                      default)))))
```

Comment:
* The `5am` package prefix is not needed since we `:use` the package
* `gen-one-element` returns a generator function, so ``(,(gen-one-element …) …)` splices the function, not a random value
* `gen-one-element` and other generators are intended to be used in conjunction with `for-all`

I suggest testing the validation behavior like this:
```lisp
(defun gen-invalid-margin-spec-element ()
  (let ((anchor (gen-one-element :relative :absolute)))
   (lambda ()
     (case (random 3)
       (0 :foo)
       (1 (list (funcall anchor) :foo))
       (2 (list (funcall anchor) 1 2))))))

(defun gen-invalid-margin-spec ()
  (let ((edge (gen-one-element :left :right :top :bottom))
        (element (gen-invalid-margin-spec-element)))
    (lambda ()
      (list (funcall edge) (funcall element)))))

(test normalize-margin-spec.validation

  (for-all ((spec (gen-invalid-margin-spec)))
    (let ((default '(:left (:relative 0)
                     :top (:relative 0)
                     :right (:relative 0)
                     :bottom (:relative 0))))
      (signals error (climi::normalize-margin-spec spec default)))))
```

Reply:
done